docs/peer-validation: add docs about peer validation (#154)
squat/kilo: docs/peer-validation.md

# Peer Validation

A [ValidatingAdmissionWebhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook) can  be used to avoid applying faulty Peer configurations to the cluster.

## How It works

A ValidatingAdmissionWebhookConfiguration specifies a resource group (Peers), a service (the actual webhook) and operations (UPDATE, CREATE).

> After applying the configuration, the Kuberntes API server is going to send an AdmissionReviewRequest to the webhook service every time a Peer gets updated or created.

Author: If you want to be generic, than update or create are not the only possible operations. I wanted to be more specific and only describe this specific configuration. This is why I wrote peer instead of resource.

> After applying the configuration, the Kuberntes API server is going to send an AdmissionReviewRequest to the webhook service every time a Peer gets updated or created.

Reviewer: OK I'm happy with either way: either keep it specific to Kilo or make it generic. However, just to make things clearer to the user, if we chose to keep it specific to Kilo, I think we should add something at the beginning of this sentence like "In the of the Kilo, ...". I say this because the sentence immediately before this one introduces the concept of ValidatingWebhookConfigurations at a high level independently from Kilo so the sentence after makes it sounds like (to me) all VWCs affect Peer resources.

> After applying the configuration, the Kuberntes API server is going to send an AdmissionReviewRequest to the webhook service every time a Peer gets updated or created.

Author: sgtm

> After applying the configuration, the Kuberntes API server is going to send an AdmissionReviewRequest to the webhook service every time a Peer gets updated or created.

Author: lets keep the first paragraph generic

> The API will only apply the requested changes to a Peer, if the Request was answered with `Allowed=true`.

Author: Same thing here. If you want to keep it generic than it is not correct to say that the resource is only then updated. You can set a failure policy for a webhook. So requests that run in a timeout are also allowed with `failurePolicy: Ignore`

> The API will only apply the requested changes to a Peer, if the Request was answered with `Allowed=true`.

Reviewer: To finish this one should we say
```suggestion
In the case of Kilo, the Kubernetes API will only apply the requested changes to a Peer, if the webhook server replies to the request with `"allowed": true`.
```

> The API will only apply the requested changes to a Peer, if the Request was answered with `Allowed=true`.

Author: i tried to make it more generic and mentioned the failurePolicy. Just pushed the changes

> The API will only apply the requested changes to a Peer, if the Request was answered with `Allowed=true`.

Author: i think i messed somthing up with the commits and rebase.

> The API will only apply the requested changes to a Peer, if the Request was answered with `Allowed=true`.

Author: OK now it's ready. I added a new paragraph to mention how the configuration would be in this specific case.

## Get started

[Kilo-Peer-Validation](https://github.com/leonnicolas/kilo-peer-validation) contains a webserver, that will reject AdmissionReviewRequest with faulty peer configurations.

Apply the Service, the deployment of the actual webserver and the ValidatingAdmissionWebhookConfiguration with
```bash
kubectl apply -f https://raw.githubusercontent.com/leonnicolas/kilo-peer-validation/main/deployment-no-cabundle.yaml
```

Since the API server will only talk to the webhook via tls, you can use [kube-webhook-certgen](https://github.com/jet/kube-webhook-certgen) to create a secret and make a certificate signing request to the Kuberntes API server.
```bash

> docker run -v /path/to/kubeconfig:/kubeconfig.yaml:ro jettech/kube-webhook-certgen:v1.5.2 --kubeconfig /kubeconfig.yaml create --namespace kilo --secret-name peer-validation-webhook-tls --host peer-validation,peer-validation.kilo.svc,peer-validation.kilo.svc.cluster.local --key-name tls.key --cert-name tls.config

Author: When i reverse lookup the service's IP address, I only get peer-validation.kilo.svc.cluster.local. Would it be useful when you are outside or in a different cluster? Well in that case you don't have the caBundle anyways.

> docker run -v /path/to/kubeconfig:/kubeconfig.yaml:ro jettech/kube-webhook-certgen:v1.5.2 --kubeconfig /kubeconfig.yaml create --namespace kilo --secret-name peer-validation-webhook-tls --host peer-validation,peer-validation.kilo.svc,peer-validation.kilo.svc.cluster.local --key-name tls.key --cert-name tls.config

Reviewer: Hmm that's a good question. This DNS name should only be used from within a cluster, ie by the API server contacting the webhook receiver. So I think we don't need to worry about the `.cluster.local` case. In the case where you _are_ running the webhookserver off-cluster hosted in some central place, you would probably want to use a different DNS name without `.cluster.local`, e.g. kilo-validation.example.com.

> docker run -v /path/to/kubeconfig:/kubeconfig.yaml:ro jettech/kube-webhook-certgen:v1.5.2 --kubeconfig /kubeconfig.yaml create --namespace kilo --secret-name peer-validation-webhook-tls --host peer-validation,peer-validation.kilo.svc,peer-validation.kilo.svc.cluster.local --key-name tls.key --cert-name tls.config

Author: I got rid of it

# Peer Validation

A [ValidatingAdmissionWebhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook) can  be used to avoid applying faulty Peer configurations to the cluster.

## How It Works

A [ValidatingWebhookConfiguration](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#configure-admission-webhooks-on-the-fly) is a Kubernetes resource that can be used to dynamically specify a service (i.e. the webhook server) that should validate operations (e.g. `UPDATE`, `CREATE`, etc.) on a particular resource (e.g. Kilo Peers).
Once such a configuration is applied, the Kubernetes API server will send an AdmissionReviewRequest to the webhook service every time the specified operations are applied to the resource of the specified type.

> With regard to the [failure policy](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#failure-policy), the API server will apply the requested changes to a resource, if the Request was answered with `"Allowed":true`, or deny the changes if the answere was `"Allowed":false`.

Reviewer: ```suggestion
With regard to the [failure policy](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#failure-policy), the API server will apply the requested changes to a resource if the request was answered with `"allowed": true`, or deny the changes if the answer was `"allowed": false`.
```

> In case of the Kilo Peer Validation, the specified operations are `UPDATE` and `CREATE`.

Reviewer: Style nit: maybe we can make this one sentence and less broken up.
```suggestion
In case of Kilo Peer Validation, the specified operations are `UPDATE` and `CREATE`, the resources are `Peers`, and the default `failurePolicy` is set to `Fail`.
```

The resources are `Peers`.

> By default the `failurePolicy` ist set to `Fail`.

Reviewer: ```suggestion
By default the `failurePolicy` is set to `Fail`.
```

> The resources are `Peers`.

Reviewer: now we have this duplicated and can remove this line and the next one :see_no_evil:

> This means that changes to Peers are not applied, if an unkown error occures when calling the webhook.

Reviewer: ```suggestion
This means that changes to Peers are not applied if an unknown error occurs when calling the webhook.
```

> View the full configuration [here](https://github.com/leonnicolas/kilo-peer-validation/blob/main/deployment-no-cabundle.yaml).

Reviewer: ```suggestion
View the full ValidatingWebhookConfiguration [here](https://github.com/leonnicolas/kilo-peer-validation/blob/main/deployment-no-cabundle.yaml).
```
